integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py

```diff
@@ -0,0 +1,201 @@
+    pipeline.connect("adapter.output", "response_llm.messages")
+
+    # Run the pipeline with a user question
+    user_input = "What is the time in New York? Be brief."
+    user_input_msg = ChatMessage.from_user(text=user_input)
+
+    result = pipeline.run({"llm": {"messages": [user_input_msg]}, "adapter": {"initial_msg": [user_input_msg]}})
+    print(result["response_llm"]["replies"][0].text)
+    ```
+
+    You can also use the toolset via MCP SSE to talk to remote servers:
+    ```python
+    from haystack_integrations.tools.mcp import MCPToolset, SSEServerInfo
+    from haystack.components.tools import ToolInvoker
+
+    # Create the toolset with an SSE connection
+    sse_toolset = MCPToolset(
+        server_info=SSEServerInfo(base_url="http://some-remote-server.com:8000"),
+    )
+
+    # Use the toolset as shown in the pipeline example above
+    ```
+    """
+
+    def __init__(self, server_info: MCPServerInfo, connection_timeout: int = 30, invocation_timeout: int = 30):
+        """
+        Initialize the MCP toolset.
+
+        :param server_info: Connection information for the MCP server
+        :param connection_timeout: Timeout in seconds for server connection
+        :param invocation_timeout: Default timeout in seconds for tool invocations
+        """
+        # Initialize with empty tools list first
+        super().__init__(tools=[])
+
+        # Store configuration
+        self.server_info = server_info
+        self.connection_timeout = connection_timeout
+        self.invocation_timeout = invocation_timeout
+
+        # Connect and load tools
+        try:
+            # Create the appropriate client using the factory method
+            client = self.server_info.create_client()
+
+            # Connect and get available tools
+            tools = self._run_sync(client.connect(), timeout=self.connection_timeout)
+
+            # Create MCPTool instances for each available tool and add them
+            for tool_info in tools:
+                tool = MCPTool(
+                    name=tool_info.name,
+                    server_info=self.server_info,
+                    description=tool_info.description,
+                    connection_timeout=self.connection_timeout,
+                    invocation_timeout=self.invocation_timeout,
+                )
+                # Handles duplicates and other validation
+                self.add(tool)
+
+        except Exception as e:
+            message = f"Failed to initialize MCPToolset: {e}"
+            raise MCPConnectionError(message=message, server_info=self.server_info, operation="initialize") from e
+
+    def _run_sync(self, coro, timeout: float | None = None):
+        """
+        Run an async coroutine synchronously.
+
+        :param coro: The coroutine to run
+        :param timeout: Optional timeout in seconds
+        :returns: The result of the coroutine
+        :raises TimeoutError: If the operation times out
+        """
+        try:
```

Comment:
Why don't simply use `asyncio.run(asyncio.wait_for(coro, timeout))` here? It will handle event loop management automatically.

Reply:
Yeah, I'll try it - there was some issue about event loop running and not running, detecting if it is and so on. I'll revisit 👍

Reply:
For sync scenarios (like this one), [asyncio.run](https://docs.python.org/3/library/asyncio-runner.html#running-an-asyncio-program) should be enough to sync-run a coroutine. For async ones, I expect you will implement a `_run_async`?

Reply:
I think I agree with @mpangrazzi here to make this simpler. We opted not to do this in AsyncPipeline currently and to avoid the `This event loop is already running` we added 
```python
import nest_asyncio
nest_asyncio.apply()
```
to the top of the Jupyter notebook we were running. See [here](https://haystack.deepset.ai/cookbook/async_pipeline).

When are you running into the `"This event loop is already running" errors`?

Reply:
Also if we really do need to keep this it seems like we have an almost duplicate implementation of `_run_sync` in `mcp_tool.py`. Could we reuse that here especially since the method could be static in both?

Reply:
Yes, will reuse it, god point!

Reply:
> 
> When are you running into the `"This event loop is already running" errors`?

I wish it was that simple - `asyncio.run` cannot be called when another asyncio event loop is running in the same thread. So when I use this approach in hayhooks - I'm running into "This event loop is already running" there. As we talked internally It is not easy to satisfy all the requirements users expect - having MCPTool work flawlessly in sync/async context, in standalone scripts, in hayhooks deployments, all without any additional explicit code on user's side.